src/ray/util/pipe_logger.h

```diff
@@ -30,71 +42,97 @@ namespace ray {
-  size_t rotation_max_file_count = 1;
-};
-
-class RotationFileHandle {
+// File handle requires active destruction via owner calling [Close].
+class RedirectionFileHandle {
  public:
-  RotationFileHandle(MEMFD_TYPE_NON_UNIQUE write_handle,
-                     std::function<void()> termination_caller)
-      : write_handle_(write_handle), termination_caller_(std::move(termination_caller)) {}
-  RotationFileHandle(const RotationFileHandle &) = delete;
-  RotationFileHandle &operator=(const RotationFileHandle &) = delete;
+  RedirectionFileHandle() = default;
+
+  // @param termination_synchronizer is used to block wait until destruction operation
+  // finishes.
+  RedirectionFileHandle(MEMFD_TYPE_NON_UNIQUE write_handle,
+                        std::function<void()> flush_fn,
+                        std::function<void()> close_fn)
+      : write_handle_(write_handle),
+        flush_fn_(std::move(flush_fn)),
+        close_fn_(std::move(close_fn)) {
+    RAY_CHECK(flush_fn_);
+    RAY_CHECK(close_fn_);
+  }
+  RedirectionFileHandle(const RedirectionFileHandle &) = delete;
+  RedirectionFileHandle &operator=(const RedirectionFileHandle &) = delete;
+  ~RedirectionFileHandle() = default;
+
+  // Synchronously flush content to storage.
+  //
+  // TODO(hjiang): Current method only flushes whatever we send to logger, but not those
+  // in the pipe; a better approach is flush pipe, send FLUSH indicator and block wait
+  // until logger sync over.
+  void Flush() {
+    RAY_CHECK(flush_fn_);
+    flush_fn_();
+  }
 
   // Used to write to.
   //
   // TODO(hjiang): I will followup with another PR to make a `FD` class, which is not
   // copiable to avoid manual `dup`.
 #if defined(__APPLE__) || defined(__linux__)
-  RotationFileHandle(RotationFileHandle &&rhs) {
+  RedirectionFileHandle(RedirectionFileHandle &&rhs) {
     write_handle_ = rhs.write_handle_;
     rhs.write_handle_ = -1;
-    termination_caller_ = std::move(rhs.termination_caller_);
+    flush_fn_ = std::move(rhs.flush_fn_);
+    close_fn_ = std::move(rhs.close_fn_);
   }
-  RotationFileHandle &operator=(RotationFileHandle &&rhs) {
+  RedirectionFileHandle &operator=(RedirectionFileHandle &&rhs) {
     if (this == &rhs) {
       return *this;
     }
     write_handle_ = rhs.write_handle_;
     rhs.write_handle_ = -1;
-    termination_caller_ = std::move(rhs.termination_caller_);
+    flush_fn_ = std::move(rhs.flush_fn_);
+    close_fn_ = std::move(rhs.close_fn_);
     return *this;
   }
-  ~RotationFileHandle() {
-    // Only invoke termination functor when handler at a valid state.
+  void Close() {
     if (write_handle_ != -1) {
-      termination_caller_();
+      close_fn_();
+      write_handle_ = -1;
     }
   }
+  void CompleteWrite(const char *data, size_t len) {
+    ssize_t bytes_written = write(write_handle_, data, len);
+    RAY_CHECK_EQ(bytes_written, static_cast<ssize_t>(len));
+  }
 
   int GetWriteHandle() const { return write_handle_; }
```

Comment:
This can be just implemented once with return type `MEMFD_TYPE_NON_UNIQUE`?

Reply:
updated